### PR TITLE
feat: support channel creation

### LINKED ISSUE #146: Support "Create Channel" operation
__Is your feature request related to a problem? Please describe.__

* Currently the Slack conenctor only allows you to send messages to channels or individuals
* Typical use-cases however also involve creation of channel. Consider:
  * Incident is reported. Then Camunda to:
    1. Create an incident channel with a custom incident number related name
    2. Send a message to that channel
    3. (Next step, not for this issue) invite people to this channel

__Describe the solution you'd like__

* Make the existing Slack connector dynamic
* Support "Create channel" operation

### Breakdown of implementation tasks

- [x] Adjust element template with a `condition` (consider existing `method`) input
  - [x] Consider input: channel name (FEEL or static)
  - [x] Consider result: _probably return created channel name, ID, and status?_
- [x] Adjust backend implementation (connector function)
  - [x] Business logic
  - [x] Automated tests (API mock)
- [ ] Adjust [documentation](https://github.com/camunda/camunda-platform-docs/pulls?q=is%3Apr+is%3Aclosed+label%3Aepic%3Aoptimize-versions+)
- [x] Consider migration case
  - [x] Ensure it works by trying out: e.g., have existing process with old template applied, and then change template under the hood => see what happens
    - [x] Learning excercise: learn about the `version` attribute in element templates which can be an alternative to an implicit migration
- [ ] Change element template in Web Modeler
- [ ] Deploy the new Connector run-time

__Describe alternatives you've considered__

* Create a seperate connector

__Additional context__

Part of camunda/team-connectors#192

### PR DESCRIPTION
**Pull request by chillleader**
## Description

Adds support of channel creation.

### WIP:
- [x] Modify event template
- [x] Business logic & tests changes 
- [x] Live test with Slack
- [x] Investigate creating private channels
- [x] Check for possible migration issues

## Related issues

closes #146

### CODE REVIEW

**Comment by tmetzke** [comment]
This looks really good already, @chillleader, great stuff!
I haven't looked at every detail yet, just one thing that came to my attention: try to be mindful of user-facing elements and the wording in there. 

For example, the `data.isPrivate` and `data.orgWide` inputs are currently hidden, but users can create their own templates and need to understand what these attributes do. If we can find more speaking names for those variables, that will certainly help people, I think. 🙂 

In general, everything in the element template is user-facing, so it should be speaking and as easy to grasp as possible. In the backend code of the Connector, you can of course use more technical terms if necessary.

**Comment by chillleader** [comment]
@tmetzke I removed these potentially unclear inputs completely for now. They don't add any value other than copying the Slack API defaults. It will make sense to support private channels when we implement support of inviting users to channel.

File: element-templates/slack-connector.json [hunk 1]
```diff
@@ -18,13 +18,21 @@
...
       "label": "Authentication"
     },
     {
       "id": "compose",
-      "label": "Compose"
+      "label": "Compose Message"
+    },
+    {
+      "id": "createChannel",
+      "label": "Create Channel"
```
Comment (tmetzke) [inline element-templates/slack-connector.json:35]: Providing good names is always the most complex task, I think. 😁 I was wondering if we could rather divide the groups into something like "Channel" and "Message". Things specific to messages (e.g. message text) could go into one group, and channel-related stuff (like channel name, description, but maybe also people to invite upon creation, being a private or public channel) into the other. What do you think?

**Comment by tmetzke** [comment]
@chillleader, can you post some images here of how the element template will look and behave with the different options? It's a bit easier to digest usually. Thank you! 🙂

File: element-templates/slack-connector.json [hunk 2]
```diff
@@ -92,6 +117,31 @@
...
+      "type": "String",
+      "feel": "optional",
+      "binding": {
+        "type": "zeebe:input",
+        "name": "data.name"
+      },
+      "constraints": {
+        "notEmpty": true,
+        "pattern": {
+          "value": "^[a-z0-9-_]{1}[a-z0-9-_]{0,20}$",
```
Comment (tmetzke) [inline element-templates/slack-connector.json:138]: Looking at the [API docs](https://api.slack.com/methods/conversations.create), I think you can use up to 80 chars for the name. I can also verify this in our Slack workspace manually. Plus, I think we can simplify the regex.
```suggestion
          "value": "^[-_a-z0-9]{1,80}$",
```

File: element-templates/slack-connector.json [hunk 3]
```diff
@@ -92,6 +117,31 @@
...
+          "message": "May contain up to 20 lowercase letters, digits, underscores and dashes"
```
Comment (tmetzke) [inline element-templates/slack-connector.json:139]:
```suggestion
          "message": "May contain up to 80 lowercase letters, digits, underscores, and dashes"
```

**Comment by tmetzke** [comment]
> @tmetzke I removed these potentially unclear inputs completely for now. They don't add any value other than copying the Slack API defaults. It will make sense to support private channels when we implement support of inviting users to channel.

I think the public/private channel creation already makes sense to add here. @MoAtCamunda, what do you think?

**Review by tmetzke — COMMENTED** [review]
The code looks really good, well done! Consider the suggestions and previous comments made here in the PR please to complete this. 🙂

File: element-templates/slack-connector.json (hunk 1 above)
Comment (chillleader) [reply element-templates/slack-connector.json:35]: I agree, I also thought about providing more obvious names. My concern was that changing an existing group name might lead to compatibility issues. On the other hand, I assume it's just a UI feature so it should work fine. Will double check it.

Comment (tmetzke) [reply element-templates/slack-connector.json:35]: I think this is just UI, yes. You can give it a try though in Desktop Modeler and Web Modeler. If you need assistance with that, let me know. 🙂

**Comment by chillleader** [comment]
**Some screenshots of how the template behaves**

Left to right:
* Initial state after creation
* After selecting create channel method
* Validation error
Screenshot 2022-11-18 at 12 53 34 Screenshot 2022-11-18 at 12 54 50 Screenshot 2022-11-18 at 12 55 14

File: element-templates/slack-connector.json (hunk 1 above)
Comment (chillleader) [reply element-templates/slack-connector.json]: Grouped the inputs as suggested (see screenshots below). Did not use a common channel name input for both methods though - for chat.postMessage method it has a slightly more complex logic and can also consume usernames/emails

**Comment by tmetzke** [comment]
But that also depends on what we maybe do in other Connectors, we should be consistent with this.

**Comment by tmetzke** [comment]
@MoAtCamunda, would be nice if you could have a look as well here to comment on the changes to the element template. 🙂

**Comment by chillleader** [comment]
> Do we expect more than public and private as scopes, or will this only be binary? Just thinking about whether a check box is easier here or not.

Maybe I'm missing something, but as per [documentation](https://docs.camunda.io/docs/components/modeler/desktop-modeler/element-templates/defining-templates/#zeebeinput) checkboxes are not supported as Zeebe inputs, are they? Pls correct me if I'm wrong

**Comment by tmetzke** [comment]
> Maybe I'm missing something, but as per [documentation](https://docs.camunda.io/docs/components/modeler/desktop-modeler/element-templates/defining-templates/#zeebeinput) checkboxes are not supported as Zeebe inputs, are they? Pls correct me if I'm wrong

Interesting, I thought I've seen them, nevermind then 👍

File: src/main/java/io/camunda/connector/slack/ConversationsCreateData.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  @NotEmpty @Secret private String newChannelName;
+  @NotNull private Visibility visibility;
+
+  @Override
+  public SlackResponse invoke(MethodsClient methodsClient) throws SlackApiException, IOException {
+    ConversationsCreateRequest request =
+        ConversationsCreateRequest.builder()
+            .name(newChannelName)
+            .isPrivate(Visibility.PRIVATE.equals(visibility))
```
[the remark this answers is not in the document]
Comment (tmetzke) [reply src/main/java/io/camunda/connector/slack/ConversationsCreateData.java:29]: Good point, classic: https://stackoverflow.com/questions/1750435/comparing-java-enum-members-or-equals

File: src/main/java/io/camunda/connector/slack/ConversationsCreateData.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        + ", visibility="
+        + visibility
+        + '}';
+  }
+
+  public enum Visibility {
+    PUBLIC,
+    PRIVATE
+  }
+}
```
[the remark this answers is not in the document]
Comment (chillleader) [reply src/main/java/io/camunda/connector/slack/ConversationsCreateData.java]: Makes sense

**Review by MoAtCamunda — APPROVED** [review]

**Pull request merged**